File: 02 백준/14 누적합/Python/Prob21921.py

"""
title  : 21921. 블로그 (Silver 3)
time   : 136ms        184ms
memory : 58784KB      172184KB
"""

import sys

def solution():

    input = sys.stdin.readline
    N, X = map(int, input().split())
    visitors = list(map(int, input().split()))

    prefix_sum = [visitors[0]]
    for idx in range(1, N):
        prefix_sum.append(prefix_sum[idx-1] + visitors[idx])

    answer = prefix_sum[X-1]
    count = 1

    for idx in range(X, N):
        value = prefix_sum[idx] - prefix_sum[idx-X]
        if answer == value:
            count += 1
        elif answer < value:
            answer = value
            count = 1

    if answer:
        print(answer, count, sep="\n")
    else:
        print("SAD")

solution()

File: 02 백준/14 누적합/Java/src/Prob21921.java

/**
* title  : 21921. 블로그 (Silver 3)
* time   : ms
* memory : KB
*/

import java.io.*;
import java.util.*;

public class Prob21921 {
    
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int X = Integer.parseInt(st.nextToken());

        int[] record = new int[N];
        
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            record[i] = Integer.parseInt(st.nextToken());
        }

        int tempValue = 0;
        for (int i = 0; i < X; i++) {
            tempValue += record[i];
        }

        int maxValue = tempValue;
        int maxCount = 0;
        for (int i = X; i < N; i++) {
            tempValue += (record[i] - record[i-X]);
            if (maxValue == tempValue) {
                maxCount++;
            } else if (maxValue < tempValue) {
                maxValue = tempValue;
                maxCount = 1;
            }
        }

        if (maxValue == 0) {
            System.out.println("SAD");
        } else {
            System.out.println(maxValue);
            System.out.println(maxCount);
        }
    }
}
